### PR TITLE
[WIP] Make equalize_channels capable of operating on Forward, Covariance and CrossSpectralDensity

### PR DESCRIPTION
**Pull request by wmvanvliet**
In the process of adding support for DICS whitening (#7021), I now find myself with an `Info`, two CSD matrices and a forward model, all of which need to have exactly the same channels defined in the same order for the computations to work. This is a common theme throughout our codebase and currently, we have a smattering of private helper functions everywhere that deal with equalizing channels one way or another.

However, we also have `mne.channels.equalize_channels`, which job it is to, given a list of instances, equalize the channels amongst them. In this PR, I've extended its capabilities to:

  1. Also deal with `Info`, `Forward`, `Covariance` and `CrossSpectralDensity`
  2. Also re-order the channels to make sure the channel orders are the same.

To accomplish this, I had to give our channel picking functionality a bit of extra love. Adding an  `ordered=True` option for those picking functions that lacked it, and adding `pick_channels` methods to `Forward`, `Covariance` and `CrossSpectralDensity`, so that `equalize_channels` doesn't need to contain a big switch for each supported instance type.

Todo:

  - [x] Add `pick_channels` methods to `Info`, `Forward`, `Covariance` and `CrossSpectralDensity`
  - [x] Add `ch_names` attribute to `Info` and `Forward`
  - [x] Add unit tests
  - [ ] ~~Replace private functions that perform channel equalization with `equalize_channels`~~ Will do in follow-up PR
  - [x] Maybe: add logic to keep objects that already have the proper order untouched for speed
  - [ ] Add "what's new" entry

Question @larsoner: why does `equalize_channels` first find the instance with the most channels defined and uses that as initial set to then intersect with the others to compute the set of common channels? Why not just pick the fist instance? Also, why is the first parameter called `candidates` and not `instances`?

### CODE REVIEW

**Comment by larsoner** [comment]
> why does equalize_channels first find the instance with the most channels defined and uses that as initial set to then intersect with the others to compute the set of common channels?

No idea / can't remember.

> Also, why is the first parameter called candidates and not instances?

It can and probably should be changed to `instances`. 

This was added by @dengemann six years ago and I actually forgot that it existed / don't use it, so maybe he has ideas.

**Comment by wmvanvliet** [comment]
I think it's main use was in `grand_average` to make sure all evokeds had the same channels.

**Comment by wmvanvliet** [comment]
I'm done now with the overhaul of the `equalize_channels` function itself. It can now equalize the channels almost anything, and new object types can be added very easily. Would you care to take a look @larsoner @dengemann?

This PR is growing a bit long. Shall I split the work of replacing the custom private functions for equalizing channels with `equalize_channels` into a different PR?

**Comment by larsoner** [comment]
> This PR is growing a bit long. Shall I split the work of replacing the custom private functions for equalizing channels with equalize_channels into a different PR?

No I think this is fine, it's not too many lines. If the tests are clear enough and CIs are happy then we should be good.

Is this ready for MRG or still WIP?

**Comment by wmvanvliet** [comment]
I'd prefer to do the refactor in another PR to keep things small. Let me update what's new and then this is ready for merge.

File: mne/channels/channels.py [hunk 1]
```diff
@@ -88,46 +88,86 @@ def _get_ch_type(inst, ch_type, allow_ref_meg=False):
 
 
 @verbose
-def equalize_channels(candidates, verbose=None):
-    """Equalize channel picks for a collection of MNE-Python objects.
+def equalize_channels(instances, copy=False, verbose=None):
```
Comment (agramfort) [inline mne/channels/channels.py]: functions should copy by default. Only methods should not do this and operated inplace

File: mne/channels/channels.py [hunk 2]
```diff
@@ -88,46 +88,86 @@ def _get_ch_type(inst, ch_type, allow_ref_meg=False):
...
+    """Equalize channel picks and ordering across multiple MNE-Python objects.
+
+    First, all channels that are not common to each object are dropped. Then,
+    using the first object in the list as a template, the channels of each
```
Comment (agramfort) [inline mne/channels/channels.py]: you change the current behavior by taking the first at template. BEfore it was taking the one with the highest number of channels so order did not matter

File: mne/tests/test_evoked.py [hunk 3]
```diff
@@ -585,7 +585,7 @@ def test_arithmetic():
     evoked2.reorder_channels(evoked2.ch_names[::-1])
     assert not np.allclose(data2, evoked2.data)
     with pytest.warns(RuntimeWarning, match='reordering'):
-        ev3 = grand_average((evoked1, evoked2))
+        ev3 = combine_evoked([evoked1, evoked2], weights=[0.5, 0.5])
```
Comment (agramfort) [inline mne/tests/test_evoked.py]: why did you need to change this?

Comment (wmvanvliet) [reply mne/tests/test_evoked.py]: because internally, `grand_average` first calls `equalize_channels` and then `combine_evoked`. In current master, `equalize_channels` does not re-order the channels, hence the re-ordering logic in `combine_evoked` is activated, which gives a warning, which is what is tested here. In this PR, `equalize_channels` does re-order the channels, without warning, because I assume re-ordering is something the user wants to do when "equalizing channels", hence the re-ordering logic in `combine_evoked` is not activates, hence no warning is issued, hence the test fails. Since the test is for the re-ordering logic of `combine_evoked`, I thought it better to just call it directly without going through `grand_average`.

File: mne/channels/channels.py (hunk 2 above)
Comment (wmvanvliet) [reply mne/channels/channels.py]: That is something I do wish to change. By using the first object as template, the user can more easily control the order in which the channels end up in. For example, it would be computionally beneficial to specify the largest object in memory first, so those channels do not need to be re-ordered which would be expensive. Using the object that originally has the most channels seems to random to me. If there was a good reason for it, I'd love to hear it (@dengemann?)

File: mne/channels/channels.py (hunk 1 above)
Comment (wmvanvliet) [reply mne/channels/channels.py]: True. However, that would be change from current master, since that version operated in-place only. I'm being inconsistent here I see, since I break backwards compatibility anyway be re-ordering channels.

I could avoid breaking backwards compatibility here by making the signature:

```python
equalize_channels(candidates, copy=False, ordering=False, verbose=None)
```

if that is preferable. However, every time I would want to use the function, I want `copy=True, ordering=True` and don't see much scenarios where you wouldn't want that.

File: mne/tests/test_evoked.py (hunk 3 above)
Comment (agramfort) [reply mne/tests/test_evoked.py]: ok

File: mne/channels/channels.py (hunk 2 above)
Comment (agramfort) [reply mne/channels/channels.py]: then it should at least be documented as an API change in what's new.

File: mne/channels/channels.py (hunk 1 above)
Comment (agramfort) [reply mne/channels/channels.py]: No strong feeling but at least API change should be documented in what's new. The current function in master operates inplace and returns modified inputs which is evil...

Comment (wmvanvliet) [reply mne/channels/channels.py]: `copy=True` it is!

**Comment by agramfort** [comment]
@wmvanvliet see why circle complains

**Comment by drammock** [comment]
```
changes/latest.inc:148: WARNING: py:attr reference target not found: mne.Forward.ch_names
changes/latest.inc:148: WARNING: py:attr reference target not found: mne.Info.ch_names
```

File: doc/changes/latest.inc [hunk 4]
```diff
@@ -132,3 +134,15 @@ API
...
+
+- :func:`mne.channels.equalize_channels` now uses the first object in the list as a template for channel ordering by `Marijn van Vliet`_
+
+- :func:`mne.channels.equalize_channels` now also re-orders the channels to match, in addition to dropping channels that are not shared by all objects by `Marijn van Vliet`_
+
+- :meth:`mne.io.Raw.pick_channels`, :meth:`mne.Epochs.pick_channels` and :meth:`mne.Evoked.pick_channels` now have an ``ordered`` parameter to enforce the ordering of the picked channels by `Marijn van Vliet`_
+
+- New methods :meth:`mne.Forward.pick_channels`, :meth:`mne.Covariance.pick_channels`, :meth:`mne.Info.pick_channels`, :meth:`mne.time_frequency.CrossSpectralDensity.pick_channels` by `Marijn van Vliet`_
+
+- New attributes :attr:`mne.Forward.ch_names`, :attr:`mne.Info.ch_names` by `Marijn van Vliet`_
```
Comment (wmvanvliet) [inline doc/changes/latest.inc]: @larsoner Could you please help me? What am I doing wrong here? Sphynx can't find these attribute references.

Comment (larsoner) [reply doc/changes/latest.inc]: These are currently not directly linkable. Either use double backticks or just link to the object:
```
attributes ``mne.Forward.ch_names``
```
or
```
attributes :class:`mne.Forward.ch_names <mne.Forward>`
```

**Comment by agramfort** [comment]
Thx @wmvanvliet
